### PR TITLE
add recreate.

### LINKED ISSUE #1916: add new option 'recreate' service
- [ ] Create a new API called `recreateService`. (update - should be a PUT request with serviceId in query). No request body)
- [ ] The service must call the existing destroy and deploy methods. (similar to migrate)
- [ ] Should not create new serviceId. 
- [ ] Force destroy the service if necessary.

### PR DESCRIPTION
**Pull request by Alice1319**
fixes #1916 

StartDestroy:
startDestroy

DestroySuccess:
destroySuccess

StartDeploy:
startDeploy

DeploySuccess:
deploySuccess

DestroyRetry:
retryDeploy

RetryDestroyManually:
retryDestroyManually

from destroy to deploy:
fromDestroy2Deploy

Retry Deploy:
retryDeploy

RetryDeployManually:
retryDeployManually

RecreateSuccessfully:
recreateSuccess

test result:
image

### CODE REVIEW

**Comment by WangLiNaruto** [comment]
LGTM

File: modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+    @Resource
+    private RecreateService recreateService;
+
+    /**
+     * Create a job to recreate the deployed service.
+     *
+     * @return response
+     */
+    @Tag(name = "Recreate", description = "APIs to manage the service recreate.")
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java:86]: we do not need a new tag here. The tag and description must be same as 

@Tag(name = "Service", description = "APIs to manage the services")

File: modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    @Operation(description = "Create a job to recreate the deployed service.")
+    @PostMapping(value = "/services/recreate", produces = MediaType.APPLICATION_JSON_VALUE)
+    @ResponseStatus(HttpStatus.ACCEPTED)
+    @AuditApiRequest(methodName = "getCspFromRequestUri")
+    public UUID recreateService(@Valid @RequestBody RecreateRequest recreateRequest) {
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java:91]: Also there should be no request body here. The recreate will just use whatever parameters that was used to deploy the service.

File: modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    private RecreateService recreateService;
+
+    /**
+     * Create a job to recreate the deployed service.
+     *
+     * @return response
+     */
+    @Tag(name = "Recreate", description = "APIs to manage the service recreate.")
+    @Operation(description = "Create a job to recreate the deployed service.")
+    @PostMapping(value = "/services/recreate", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java:88]: should be PUT with serviceId in the URI.

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/recreate/steps/ProcessRecreateDeploymentResult.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    @Override
+    public void execute(DelegateExecution execution) {
+
+        String processInstanceId = execution.getProcessInstanceId();
+        Map<String, Object> variables = runtimeService.getVariables(processInstanceId);
+        UUID originalServiceId = (UUID) variables.get(RecreateConstants.ID);
+        log.info("Recreate workflow of Instance Id : {} check deploy service status with id:{}",
+                processInstanceId, originalServiceId.toString());
```
Comment (swaroopar) [inline modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/recreate/steps/ProcessRecreateDeploymentResult.java:58]: i have not checked this in detail. But pls ensure that the same entry in DEPLOY_SERVICE table is used to destroy and also deploy again.

**Review by swaroopar — CHANGES_REQUESTED** [review]

File: modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Alice1319) [reply modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java:91]: we will fix this one at another userstory.

File: modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+     * Create a job to recreate the deployed service.
+     *
+     * @return response
+     */
+    @Tag(name = "Service", description = "APIs to manage the services")
+    @Operation(description = "Create a job to recreate the deployed service.")
+    @PutMapping(value = "/services/recreate/{serviceId}", produces =
+            MediaType.APPLICATION_JSON_VALUE)
+    @ResponseStatus(HttpStatus.ACCEPTED)
+    @AuditApiRequest(methodName = "getCspFromRequestUri")
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java:86]: this must be `@AuditApiRequest(methodName = "getCspFromServiceId")`

File: modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+     *
+     * @param recreateId ID of the service recreate.
+     * @return serviceRecreateEntity.
+     */
+    @Tag(name = "Service", description = "APIs to manage the services")
+    @Operation(description = "Get recreate records based on recreate id.")
+    @GetMapping(value = "/services/recreate/{recreateId}", produces =
+            MediaType.APPLICATION_JSON_VALUE)
+    @ResponseStatus(HttpStatus.OK)
+    @AuditApiRequest(methodName = "getCspFromServiceRecreateId")
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/controllers/ServiceRecreateApi.java:166]: where is this method 'getCspFromServiceRecreateId' created?

**Review by swaroopar — CHANGES_REQUESTED** [review]

**Review by swaroopar — APPROVED** [review]
LGTM

**Pull request merged**